File: src/lib/signals/strategies/mean-reversion-bb.ts
import { atr, bollinger, closes, stddev, sma } from "../indicators";
import type { Strategy, StrategyContext, StrategySignal } from "../types";

const DEFAULTS = {
  bbPeriod: 20,
  bbStd: 2,
  atrPeriod: 14,
  atrStopMult: 1.5,
  entryZ: 2.0,
  exitZ: 0.25,
};

/**
 * Bollinger-band / z-score mean reversion.
 * Goes long when price is >entryZ stddevs below the moving average;
 * short when >entryZ above. Exits to flat when |z| drops below exitZ
 * (price has returned toward the mean).
 *
 * Stops are ATR-based; target is the middle band.
 */
export function meanReversionBB(overrides: Partial<typeof DEFAULTS> = {}): Strategy {
  const params = { ...DEFAULTS, ...overrides };
  const warmup = Math.max(params.bbPeriod, params.atrPeriod) + 2;

  return {
    id: "mean-reversion-bb",
    params,
    warmupBars: warmup,
    onCandle(ctx: StrategyContext): StrategySignal | null {
      const candles = ctx.candles.toArray();
      if (candles.length < warmup) return null;

      const c = closes(candles);
      const { bbPeriod, bbStd, atrPeriod, atrStopMult, entryZ, exitZ } = params;

      const bb = ctx.indicators.get(`bb:${bbPeriod}:${bbStd}`, () =>
        bollinger(c, bbPeriod, bbStd),
      );
      const sd = ctx.indicators.get(`sd:${bbPeriod}`, () => stddev(c, bbPeriod));
      const m = ctx.indicators.get(`sma:${bbPeriod}`, () => sma(c, bbPeriod));
      const a = ctx.indicators.get(`atr:${atrPeriod}`, () => atr(candles, atrPeriod));

      const i = candles.length - 1;
      const px = c[i];
      const mid = m[i];
      const sigma = sd[i];
      const atrV = a[i];
      if (!Number.isFinite(mid) || !Number.isFinite(sigma) || sigma === 0 || !Number.isFinite(atrV)) {
        return null;
      }

      const z = (px - mid) / sigma;
      const features: Record<string, number> = {
        zscore: z,
        mid,
        upper: bb.upper[i],
        lower: bb.lower[i],
        atr: atrV,
      };

      // Exit: we've returned near the mean
      if (Math.abs(z) < exitZ) {
        return mkSignal("flat", 0, Math.min(1, 1 - Math.abs(z) / exitZ), ctx.now, features);
      }
      // Entry long: oversold
      if (z <= -entryZ) {
        const stop = px - atrStopMult * atrV;
        return mkSignal("long", clamp(-z / 3, -1, 1), clamp(Math.abs(z) / (entryZ * 1.5), 0, 1), ctx.now, features, {
          price: px,
          stop,
          target: mid,
        });
      }
      // Entry short: overbought
      if (z >= entryZ) {
        const stop = px + atrStopMult * atrV;
        return mkSignal("short", clamp(-z / 3, -1, 1), clamp(Math.abs(z) / (entryZ * 1.5), 0, 1), ctx.now, features, {
          price: px,
          stop,
          target: mid,
        });
      }
      return null;
    },
  };
}

function mkSignal(
  side: StrategySignal["side"],
  score: number,
  confidence: number,
  ts: number,
  features: Record<string, number>,
  entryHint?: StrategySignal["entryHint"],
): StrategySignal {
  return { strategyId: "mean-reversion-bb", ts, side, score, confidence, features, entryHint };
}

function clamp(x: number, lo: number, hi: number): number {
  return Math.max(lo, Math.min(hi, x));
}

File: python/backtester/strategies/mean_reversion_bb.py
"""Mirror of src/lib/signals/strategies/mean-reversion-bb.ts."""
from dataclasses import dataclass, field
from typing import Any

import numpy as np

from ..indicators import atr, bollinger, sma, stddev


@dataclass
class MeanReversionBB:
    bb_period: int = 20
    bb_std: float = 2.0
    atr_period: int = 14
    atr_stop_mult: float = 1.5
    entry_z: float = 2.0
    exit_z: float = 0.25
    params: dict[str, Any] = field(default_factory=dict)

    id: str = "mean-reversion-bb"

    @property
    def warmup_bars(self) -> int:
        return max(self.bb_period, self.atr_period) + 2

    def evaluate(self, h, l, c):
        """Vectorized evaluation: returns a DataFrame-like dict of
        arrays (side, score, confidence, zscore, mid, stop, target) of the
        same length as the inputs, with NaN sides during warmup."""
        n = len(c)
        m = sma(c, self.bb_period)
        s = stddev(c, self.bb_period)
        bb = bollinger(c, self.bb_period, self.bb_std)
        a = atr(h, l, c, self.atr_period)

        with np.errstate(invalid="ignore", divide="ignore"):
            z = np.where(s > 0, (c - m) / s, np.nan)

        side = np.full(n, "null", dtype=object)
        score = np.full(n, np.nan)
        confidence = np.full(n, np.nan)
        stops = np.full(n, np.nan)
        targets = np.full(n, np.nan)

        valid = ~(np.isnan(m) | np.isnan(s) | (s == 0) | np.isnan(a))
        for i in np.where(valid)[0]:
            zi = z[i]
            if abs(zi) < self.exit_z:
                side[i] = "flat"
                score[i] = 0.0
                confidence[i] = min(1.0, 1 - abs(zi) / self.exit_z)
            elif zi <= -self.entry_z:
                side[i] = "long"
                score[i] = max(-1.0, min(1.0, -zi / 3))
                confidence[i] = max(0.0, min(1.0, abs(zi) / (self.entry_z * 1.5)))
                stops[i] = c[i] - self.atr_stop_mult * a[i]
                targets[i] = m[i]
            elif zi >= self.entry_z:
                side[i] = "short"
                score[i] = max(-1.0, min(1.0, -zi / 3))
                confidence[i] = max(0.0, min(1.0, abs(zi) / (self.entry_z * 1.5)))
                stops[i] = c[i] + self.atr_stop_mult * a[i]
                targets[i] = m[i]

        return {
            "side": side,
            "score": score,
            "confidence": confidence,
            "zscore": z,
            "middle": m,
            "upper": bb["upper"],
            "lower": bb["lower"],
            "atr": a,
            "stop": stops,
            "target": targets,
        }
